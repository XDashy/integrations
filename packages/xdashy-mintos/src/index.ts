import { unformat } from 'accounting';
import { load } from 'cheerio';
import * as request from 'superagent';

interface IMintosOptions {
  credentials: {
    username: string;
    password: string;
  };
}

export interface IProfileOverview {
  available_funds: number;
  invested_funds: number;
  total: number;
}

export interface IBalance {
  net_annual_return: number;
  interest: number;
  late_payment_fees: number;
  bad_debt: number;
  secondary_market_transactions: number;
  service_fees: number;
  campaign_rewards: number;
  total_profit: number;
}

export interface IOverview {
  balance: IBalance;
  overview: IProfileOverview;
}

const BASE_URL = 'https://www.mintos.com';
const LOGIN_URL = `${BASE_URL}/en/login/check`;
const OVERVIEW_URL = `${BASE_URL}/en/overview`;

export const parseCSRFToken = (document: string): string => {
  const $ = load(document);
  const inputEl$ = $('#login-form input[name="_csrf_token"]').first();

  return inputEl$.val();
};

const cleanNumber = (element: Cheerio): number =>
  unformat(element.text().trim());

export const parseOverviewPage = (document: string): IProfileOverview => {
  const $ = load(document);
  const balanceBoxEl$ = $('#mintos-boxes').children().eq(0);

  return {
    available_funds: cleanNumber(
      balanceBoxEl$.find('tr:nth-child(1) > td:nth-child(2)')
    ),
    invested_funds: cleanNumber(
      balanceBoxEl$.find('tr:nth-child(2) > td:nth-child(2)')
    ),
    total: cleanNumber(
      balanceBoxEl$.find('tr:nth-child(3) > td:nth-child(2)')
    ),
  };
};

export const parseBalanceOverview = (document: string) => {
  const $ = load(document);
  const box$ = $('#mintos-boxes').children().eq(1);

  return {
    net_annual_return: cleanNumber(box$.find('div.header > div')),
    interest: cleanNumber(box$.find('tr:nth-child(1) > td:nth-child(2)')),
    late_payment_fees: cleanNumber(box$.find('tr:nth-child(2) > td:nth-child(2)')),
    bad_debt: cleanNumber(box$.find('tr:nth-child(3) > td:nth-child(2)')),
    secondary_market_transactions: cleanNumber(box$.find('tr:nth-child(4) > td:nth-child(2)')),
    service_fees: cleanNumber(box$.find('tr:nth-child(5) > td:nth-child(2)')),
    campaign_rewards: cleanNumber(box$.find('tr:nth-child(6) > td:nth-child(2)')),
    total_profit: cleanNumber(box$.find('tr.em > td:nth-child(2)')),
  };
};

const getOverview = (document: string): IOverview => ({
  overview: parseOverviewPage(document),
  balance: parseBalanceOverview(document),
});

export const createMintosScraper = (options: IMintosOptions) => {
  const { credentials: { username, password } } = options;

  return {
    scrape: async () => {
      // Get the CSRF from login page
      const agent = request.agent();

      const loginPage = await agent.get(BASE_URL);
      const csrfToken = parseCSRFToken(loginPage.text);

      const formInput = {
        '_csrf_token': csrfToken,
        '_username': username,
        '_password': password,
      };

      await agent.post(LOGIN_URL)
        .type('form')
        .send(formInput);

      const overviewPage = await agent.get(OVERVIEW_URL);
      const overviewPageContent = overviewPage.text;

      return getOverview(overviewPageContent);
    },
  };
};
