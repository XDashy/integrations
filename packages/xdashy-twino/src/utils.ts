import t, {
  Type, array, failure, success, number, undefined as Undefined, union,
} from 'io-ts';

export function maybe<RT extends t.Any>(
  type: RT,
  name?: string
): t.UnionType<[RT, t.UndefinedType], t.TypeOf<RT> | undefined, t.OutputOf<RT> | undefined, t.InputOf<RT> | undefined> {
  return union<[RT, t.UndefinedType]>([type, Undefined], name)
}

export const DateFromArray = new Type<Date, Array<number>>(
  'DateFromArray',
  (v): v is Date => v instanceof Date,
  (s, c) => array(number).validate(s, c).chain(n => {
      return (n.length !== 7)
        ? failure(n, c)
        : success(new Date(n[0], n[1] - 1, n[2], n[3], n[4], n[5], n[6]));
    }),
  (a) => [
    a.getFullYear(), a.getMonth(), a.getDay(),
    a.getHours(), a.getMinutes(), a.getSeconds(), a.getMilliseconds()
  ],
);

