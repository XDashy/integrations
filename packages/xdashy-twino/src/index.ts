import * as request from 'superagent';
import { URL } from 'url';
import { ThrowReporter } from 'io-ts/lib/ThrowReporter';
import t, { type, number, exact } from 'io-ts';
import { DateFromArray } from './utils';

export const BASE_URL = new URL('https://www.twino.eu');

export const LOGIN_URL = new URL('/ws/public/login', BASE_URL);
export const ACCOUNT_SUMMARY_URL = new URL('/ws/web/investor/my-account-summary', BASE_URL);

interface TwinoOptions {
  credentials: {
    username: string;
    password: string;
  };
}

interface ProfileOverview {
  deposits: number;
  withdrawals: number;
  xirr: number;
  investment_balance: number;
  loss_on_writeoff: number;
  penalties: number;
  currency_fluctuations: number;
  account_value: number;
  investments: number;
  income: number;
  interest_without_currency_fluctuations: number;
  first_buy_share_date: Date;
}

interface Integration {
  user_profile: ProfileOverview;
  integration: {
    base_url: string;
    overview_url: string;
  },
}

const ProfileOverviewResponse = type({
  deposits: number,
  withdrawals: number,
  xirr: number,
  investmentBalance: number,
  lossOnWriteoff: number,
  penalties: number,
  currencyFluctuations: number,
  accountValue: number,
  investments: number,
  income: number,
  interestWithoutCurrencyFluctuations: number,
  firstBuyShareDate: DateFromArray,
});

const ProfileOverviewExact = exact(ProfileOverviewResponse);
export interface IProfileOverview extends t.TypeOf<typeof ProfileOverviewExact> {}

export const createTwinoScraper = (options: TwinoOptions) => {
  const { credentials } = options;

  const transformOutput = (profile: IProfileOverview) => ({
    deposits: profile.deposits,
    withdrawals: profile.withdrawals,
    xirr: profile.xirr,
    investment_balance: profile.investmentBalance,
    loss_on_writeoff: profile.lossOnWriteoff,
    currency_fluctuations: profile.currencyFluctuations,
    account_value: profile.accountValue,
    investments: profile.investments,
    income: profile.income,
    interest_without_currency_fluctuations: profile.interestWithoutCurrencyFluctuations,
    first_buy_share_date: profile.firstBuyShareDate,
  } as ProfileOverview);

  return {
    scrape: async (): Promise<Integration> => {
      const agent = request.agent();

      const formInput = {
        name: credentials.username,
        password: credentials.password,
      };
      await agent
        .post(LOGIN_URL.toString())
        .send(formInput);

      const summaryPage = await agent.get(ACCOUNT_SUMMARY_URL.toString());

      const decodeResult = ProfileOverviewExact.decode(summaryPage.body);
      ThrowReporter.report(decodeResult);

      if (decodeResult.isLeft()) {
        throw new Error('Failed to decode Twino profile overview response');
      }

      return {
        user_profile: transformOutput(decodeResult.value),
        integration: {
          base_url: BASE_URL.toString(),
          overview_url: ACCOUNT_SUMMARY_URL.toString(),
        },
      };
    },
  };
};
