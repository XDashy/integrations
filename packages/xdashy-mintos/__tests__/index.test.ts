import * as nock from 'nock';
import * as fs from 'fs';
import { createMintosScraper, parseOverviewPage, parseBalanceOverview } from '../src';

describe('Mintos', () => {

  it('login and get the user profile overview', async () => {
    nock('https://www.mintos.com')
      .get('/').replyWithFile(200, `${__dirname}/http/base-page.html`)
      .post('/en/login/check').reply(200)
      .get('/en/overview').replyWithFile(200, `${__dirname}/http/overview-page.html`);

    const credentials = {
      username: 'myUsername',
      password: 'myPassword',
    };
    const mintos = createMintosScraper({ credentials });
    const overview = await mintos.scrape();

    expect(overview.balance).toMatchSnapshot();
    expect(overview.overview).toMatchSnapshot();
  });
});

describe('parsing profile overview', () => {
  const document = fs.readFileSync(`${__dirname}/http/overview-block.html`, 'utf-8');

  it('parse success case', () => {
    expect(parseOverviewPage(document)).toMatchSnapshot();
  });

  it('parse net annual return overview', () => {
    expect(parseBalanceOverview(document)).toMatchSnapshot();
  });
});
