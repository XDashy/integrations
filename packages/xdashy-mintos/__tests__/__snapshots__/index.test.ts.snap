// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mintos login and get the user profile overview 1`] = `
Object {
  "bad_debt": -9.1,
  "campaign_rewards": 0,
  "interest": 300.88,
  "late_payment_fees": 0.89,
  "net_annual_return": 11.3,
  "secondary_market_transactions": 0,
  "service_fees": 0,
  "total_profit": 232.21,
}
`;

exports[`Mintos login and get the user profile overview 2`] = `
Object {
  "available_funds": 0.01,
  "invested_funds": 2012.2,
  "total": 2012.21,
}
`;

exports[`parsing profile overview parse net annual return overview 1`] = `
Object {
  "bad_debt": -6.93,
  "campaign_rewards": 45,
  "interest": 237.63,
  "late_payment_fees": 0.3,
  "net_annual_return": 10.11,
  "secondary_market_transactions": 0,
  "service_fees": 0,
  "total_profit": 231,
}
`;

exports[`parsing profile overview parse success case 1`] = `
Object {
  "available_funds": 0.01,
  "invested_funds": 1510.99,
  "total": 1511,
}
`;
